Add new Add Torrent experience to WebUI

Piccirello opened the pull request:
This PR uses the new APIs from #21015 to provide a WebUI Add Torrent experience more closely matching the GUI's.

This PR will be rebased and taken out of draft mode once #21015 is merged.

Before:
![before](https://github.com/user-attachments/assets/52024ea6-acbf-4153-bc4a-7315ae1736ff)

After:
![after](https://github.com/user-attachments/assets/503346e0-0116-4ec5-97f9-883dcd269f3f)

GUI, for comparison:
![gui](https://github.com/user-attachments/assets/04ba2940-55af-465f-bcff-af956a994f1c)

zhen-huan-hu commented:
Anyway we can add navigation of the remote file system, or auto-complete the remote path?

Piccirello commented on src/webui/www/private/scripts/dynamicTable.js:
```diff
@@ -2726,7 +2726,7 @@ window.qBittorrent.DynamicTable ??= (() => {
             if (node.isFolder) {
                 if (!this.collapseState.has(node.rowId))
-                    this.collapseState.set(node.rowId, { depth: depth, collapsed: depth > 0 });
+                    this.collapseState.set(node.rowId, { depth: depth, collapsed: false });
```
We achieve the desired behavior by manually collapsing all nodes via `torrentFilesTable.collapseAllNodes()`. We could really have either default when adding to the table - collapse all or expand all. I don't have a strong opinion either way and opted to keep this default since we already have other code that calls `collapseAllNodes()` after constructing the table.

glassez commented:
> > I would prefer all independent commits to be extracted into separate PRs.
> 
> I've split off several commits into #22879, #22880, and #22881.

I was talking about independent commits, i.e. not related to `WebUI: Add new Add Torrent experience` topic.
From what you've extracted, only the first one matches that.
I also meant `Reorder menu options to match toolbar` and `Switch from Object to Set`.

thalieht commented:
1. When grabbing the left/right edge of the window to resize it, it resizes normally when i expand it but it doesn't when i shrink it and when i move the cursor outside the window, it's as if i never let go of the grab.
2. If i try to add a torrent that already exists, the window still shows up normally and just merges trackers, no indication that the torrent already exists except if i read the log. I know there is no "Ask to merge trackers for manually added torrents" in WebUI but now there are a lot of options in the dialog and people may waste their time changing them for nothing.
3. Selecting a category from the drop-down doesn't fill the field with it nor does it actually assign it to the torrent. Typing it manually works normally.
4. I don't know if save path fields can be...not fixed length but current length seems uncomfortably small.
5. Auto fill suggestions are almost useless (unless you select them to see what they suggest) if the path is long enough because they are elided to fit the save path's field width.
6. `Use another path for incomplete torrent` is unchecked but the global path is still used. In the GUI if that option is unchecked it doesn't use an incomplete path at all and when it is checked it is filled with the global incomplete path initially.
7. I can never edit the save path if default `Torrent Management Mode` is `Automatic` even when i change it to `Manual`. Incomplete path doesn't have this problem. On the other hand, if `Manual` is the default, the save path field stays editable (but not used thankfully) even when i change TMM to `Automatic`.
8. With `Don't create subfolder` the added torrent's content tab only shows the contents of a folder inside the torrent. All other files which were directly in the root folder don't show up. Everything looks normal in the GUI's content tab for that torrent. If i change the priority of any files in that folder in this new window, it's not applied when the torrent is added, verified by looking at the GUI. Priority changes to the files in the root folder that don't show up after adding the torrent _are_ applied.
9. `Filter files...` is erratic. Sometimes it includes irrelevant files/folders. I can't see a pattern. 
10. The tree isn't updated when changing `Content layout` (to remove/add the root folder). I guess this one's not important.

Chocobo1 commented on src/webui/www/private/scripts/dynamicTable.js:
```diff
@@ -66,6 +67,9 @@ window.qBittorrent.DynamicTable ??= (() => {
 
     let DynamicTableHeaderContextMenuClass = null;
 
+    if (typeof LocalPreferences === "undefined")
+        window.LocalPreferences = new window.qBittorrent.LocalPreferences.LocalPreferences();
```
I think this should be a local variable. It is quite messy to inject into global IMO.

Piccirello replied:
This file already assumes there is a declared and initialized variable named `LocalPreferences`, which effectively means it expects this global variable to exist. Without this change, the table fails to initialize in the Add Torrent frame because `LocalPreferences` isn't defined.

Chocobo1 replied:
>This file already assumes there is a declared and initialized variable named LocalPreferences, which effectively means it expects this global variable to exist.

Right.
However, taking a step back. It seems to me that `LocalPreferences` don't really require to be global. I mean it should be preferred to be local.
If this is too complex to change in this PR (which I hope it is not) then it can be deferred later.

Chocobo1 commented on src/webui/www/private/scripts/mocha-init.js:
```diff
@@ -283,31 +283,30 @@ const initializeWindows = () => {
-    addClickEvent("upload", (e) => {
-        e.preventDefault();
-        e.stopPropagation();
+    document.querySelector("#uploadButton #fileselectButton").addEventListener("click", function() {
```
```suggestion
    document.querySelector("#uploadButton #fileselectButton").addEventListener("click", function(event) {
```

Chocobo1 commented on src/webui/www/private/scripts/mocha-init.js:
```diff
@@ -283,31 +283,30 @@ const initializeWindows = () => {
+        }
     });
 
+    document.querySelectorAll("#uploadButton #fileselectButton, #uploadLink #fileselectLink").forEach((element) => element.addEventListener("change", () => {
+        if (element.files.length === 0)
+            return;
+
+        window.qBittorrent.Client.uploadTorrentFiles(element.files);
+    }));
+
```
I would suggest using a for loop instead. Mostly for consistency reason.
```suggestion
    for (const element of document.querySelectorAll("#uploadButton #fileselectButton, #uploadLink #fileselectLink")) {
        element.addEventListener("change", (event) => {
            if (element.files.length === 0)
                return;

            window.qBittorrent.Client.uploadTorrentFiles(element.files);
        }
    }

```

Chocobo1 commented on src/webui/www/private/scripts/client.js:
```diff
@@ -121,6 +123,90 @@ window.qBittorrent.Client ??= (() => {
+            formData.append("file", file);
+        }
+
+        const xhr = new XMLHttpRequest();
```
We use `fetch` api now.

Chocobo1 commented on src/webui/www/private/scripts/addtorrent.js:
```diff
@@ -0,0 +1,291 @@
+    let tags = [];
+    let defaultSavePath = "";
+    let windowId = "";
+    let source;
```
```suggestion
    let source = "";
```

Chocobo1 commented on src/webui/www/private/scripts/addtorrent.js:
```diff
@@ -0,0 +1,291 @@
+        getTags();
+    });
+
+    window.addEventListener("DOMContentLoaded", () => {
```
```suggestion
    window.addEventListener("DOMContentLoaded", (event) => {
```

Chocobo1 commented on src/webui/www/private/scripts/torrent-content.js:
```diff
@@ -0,0 +1,572 @@
+    const FilePriority = window.qBittorrent.FileTree.FilePriority;
+    const TriState = window.qBittorrent.FileTree.TriState;
+    let torrentFilesFilterInputTimer = -1;
+    let onFilePriorityChanged;
```
Can you initialize it with something?

Chocobo1 commented on src/webui/www/private/scripts/torrent-content.js:
```diff
@@ -0,0 +1,572 @@
+            fileIds.push(node.data.fileId);
+
+            if (node.isFolder) {
+                node.children.forEach((child) => {
```
Same for-loop suggestion. You could do it if you want. Not a hard requirement since this seems to be moved old code.

Piccirello replied:
Going to leave this code as-is since it's unchanged

Piccirello commented:
> 1. When grabbing the left/right edge of the window to resize it, it resizes normally when i expand it but it doesn't when i shrink it and when i move the cursor outside the window, it's as if i never let go of the grab.
> 2. If i try to add a torrent that already exists, the window still shows up normally and just merges trackers, no indication that the torrent already exists except if i read the log. I know there is no "Ask to merge trackers for manually added torrents" in WebUI but now there are a lot of options in the dialog and people may waste their time changing them for nothing.
> 3. Selecting a category from the drop-down doesn't fill the field with it nor does it actually assign it to the torrent. Typing it manually works normally.
> 4. I don't know if save path fields can be...not fixed length but current length seems uncomfortably small.
> 5. Auto fill suggestions are almost useless (unless you select them to see what they suggest) if the path is long enough because they are elided to fit the save path's field width.
> 6. `Use another path for incomplete torrent` is unchecked but the global path is still used. In the GUI if that option is unchecked it doesn't use an incomplete path at all and when it is checked it is filled with the global incomplete path initially.
> 7. I can never edit the save path if default `Torrent Management Mode` is `Automatic` even when i change it to `Manual`. Incomplete path doesn't have this problem. On the other hand, if `Manual` is the default, the save path field stays editable (but not used thankfully) even when i change TMM to `Automatic`.
> 8. With `Don't create subfolder` the added torrent's content tab only shows the contents of a folder inside the torrent. All other files which were directly in the root folder don't show up. Everything looks normal in the GUI's content tab for that torrent. If i change the priority of any files in that folder in this new window, it's not applied when the torrent is added, verified by looking at the GUI. Priority changes to the files in the root folder that don't show up after adding the torrent _are_ applied.
> 9. `Filter files...` is erratic. Sometimes it includes irrelevant files/folders. I can't see a pattern.
> 10. The tree isn't updated when changing `Content layout` (to remove/add the root folder). I guess this one's not important.

Thank you for the thorough testing! Will reproduce these scenarios and put up fixes shortly.

> The point of extracting independent commits is also that everything that can be regarded as an independent fix/improvement can be reviewed and merged faster, without waiting for the review of the complex PR to be completed (which can take quite a long time in some cases). In addition, some of these fixes may even be candidates for backporting in some cases.

That makes sense. This change is super trivial and not worthy of a backport, so I may just squash it into an existing commit.

Piccirello commented on src/webui/www/private/download.html:
```diff
@@ -3,190 +3,60 @@
         "use strict";
 
         window.addEventListener("DOMContentLoaded", (event) => {
+            document.getElementById("urls").focus();
```
I'm not sure it really makes any difference, and existing code also seems to do it at the top.

Chocobo1 replied:
The intuition is that the focus *might* be lost if JS dynamically add/remove elements after this initial `focus()`.
I'm not sure how the browser will enforce this initial `focus()` and if you think it is fine then you can leave it as is.

Piccirello commented:
> 1. When grabbing the left/right edge of the window to resize it, it resizes normally when i expand it but it doesn't when i shrink it and when i move the cursor outside the window, it's as if i never let go of the grab.

I believe what you're experiencing when shrinking the window is that the table has a minimum width. Is that right? Also are you able to reproduce the window dragging after letting go with other windows (e.g. Statistics)? I don't believe this behavior is unique to the Add Torrent window.

> 2. If i try to add a torrent that already exists, the window still shows up normally and just merges trackers, no indication that the torrent already exists except if i read the log. I know there is no "Ask to merge trackers for manually added torrents" in WebUI but now there are a lot of options in the dialog and people may waste their time changing them for nothing.

This isn't a new bug/issue but I'd be open to brainstorming how we can make this better. I think fixing it should be a follow up PR since it's not new.

> 3. Selecting a category from the drop-down doesn't fill the field with it nor does it actually assign it to the torrent. Typing it manually works normally.

Fixed

> 4. I don't know if save path fields can be...not fixed length but current length seems uncomfortably small.

I've changed these fields to auto expand to the full available space. This should help as expanding the window will make the fields larger.

> 5. Auto fill suggestions are almost useless (unless you select them to see what they suggest) if the path is long enough because they are elided to fit the save path's field width.

Changes from the previous item should help with this.

> 6. `Use another path for incomplete torrent` is unchecked but the global path is still used. In the GUI if that option is unchecked it doesn't use an incomplete path at all and when it is checked it is filled with the global incomplete path initially.

I believe this is now fixed - could you try this again?

> 7. I can never edit the save path if default `Torrent Management Mode` is `Automatic` even when i change it to `Manual`. Incomplete path doesn't have this problem. On the other hand, if `Manual` is the default, the save path field stays editable (but not used thankfully) even when i change TMM to `Automatic`.

Fixed

> 8. With `Don't create subfolder` the added torrent's content tab only shows the contents of a folder inside the torrent. All other files which were directly in the root folder don't show up. Everything looks normal in the GUI's content tab for that torrent. If i change the priority of any files in that folder in this new window, it's not applied when the torrent is added, verified by looking at the GUI. Priority changes to the files in the root folder that don't show up after adding the torrent _are_ applied.

Do you have a torrent file/magnet URI you should share to help reproduce? Or a gif you could post of this behavior.

> 9. `Filter files...` is erratic. Sometimes it includes irrelevant files/folders. I can't see a pattern.

I'm not able to reproduce. This behavior is the same as the Content tab - could you confirm (by adding the torrent and then filtering there)?

> 10. The tree isn't updated when changing `Content layout` (to remove/add the root folder). I guess this one's not important.

Good catch, I didn't realize the GUI has this behavior. Thinking this through, I think there's quite a bit of complexity involved, so I would recommend holding this as a potential future improvement.

thalieht commented:
>I believe what you're experiencing when shrinking the window is that the table has a minimum width. Is that right?

If you mean the initial effect then yes, kind of, but after i bring cursor out of the window then i can shrink it more than what it started with.
>Also are you able to reproduce the window dragging after letting go with other windows (e.g. Statistics)?

I can't reproduce with other windows except for this window and `Add Torrent File...`. However i can reproduce this in master and 5.1.0.
>>3. Selecting a category from the drop-down doesn't fill the field with it nor does it actually assign it to the torrent. Typing it manually works normally.
>
>Fixed

Now i see another bug: When a category is already chosen, changing from Manual to Automatic doesn't fill the category's save path, it makes it blank.

>>4. I don't know if save path fields can be...not fixed length but current length seems uncomfortably small.
>
>I've changed these fields to auto expand to the full available space. This should help as expanding the window will make the fields larger.

:+1: 
Is it possible to left align the TMM combobox and Save Path field? Right alignment doesn't expand the Save Path length much.
>>Use another path for incomplete torrent is unchecked but the global path is still used. In the GUI if that option is unchecked it doesn't use an incomplete path at all and when it is checked it is filled with the global incomplete path initially.
>
>I believe this is now fixed - could you try this again?

Not fixed. Not sure if this matters when testing this: I can't add a torrent which has downloaded the metadata in this window anymore.

>>8. With Don't create subfolder the added torrent's content tab only shows the contents of a folder inside the torrent. All other files [...]
>
>Do you have a torrent file/magnet URI you should share to help reproduce? Or a gif you could post of this behavior.

Can't do it again because i can't add a torrent with downloaded metadata in this window anymore.
>>9. Filter files... is erratic. Sometimes it includes irrelevant files/folders. I can't see a pattern.
>
>I'm not able to reproduce. This behavior is the same as the Content tab - could you confirm?

I don't remember how i did it but i can't reproduce anymore and yes it's the same with Content tab.

Piccirello commented:
> If you mean the initial effect then yes, kind of, but after i bring cursor out of the window then i can shrink it more than what it started with.

The add torrent files table has an effective minimum width of ~455px, which is what you're experiencing. The left panes are more dynamic. I haven't tested this yet on mobile (realizing now I should try that) but if it's an issue we may be able to get creative, like collapsing the two panes into a single pane.

> 
> > Also are you able to reproduce the window dragging after letting go with other windows (e.g. Statistics)?
> 
> I can't reproduce with other windows except for this window and `Add Torrent File...`. However i can reproduce this in master and 5.1.0.

Sorry, you lost me. You said you can't reproduce this with other windows when using this branch, but are you able to reproduce it with other windows when using `master`? I think I misunderstood.

> Now i see another bug: When a category is already chosen, changing from Manual to Automatic doesn't fill the category's save path, it makes it blank.

Nice catch - this is now fixed.

> 👍 Is it possible to left align the TMM combobox and Save Path field? Right alignment doesn't expand the Save Path length much.

Sure thing, I've now applied this to the "Save at" and "Torrent settings" sections. "Torrent information" already exhibited this behavior.

> > > 8. With Don't create subfolder the added torrent's content tab only shows the contents of a folder inside the torrent. All other files [...]
> > 
> > 
> > Do you have a torrent file/magnet URI you should share to help reproduce? Or a gif you could post of this behavior.
> 
> Can't do it again because i can't add a torrent with downloaded metadata in this window anymore.

Can you retest with a non-hybrid torrent?

> > > 9. Filter files... is erratic. Sometimes it includes irrelevant files/folders. I can't see a pattern.
> > 
> > 
> > I'm not able to reproduce. This behavior is the same as the Content tab - could you confirm?
> 
> I don't remember how i did it but i can't reproduce anymore and yes it's the same with Content tab.

Could it be that a folder is unexpectedly getting included because it contains a file with a name that matches the search? I've been thrown off by that before.

> > > Use another path for incomplete torrent is unchecked but the global path is still used. In the GUI if that option is unchecked it doesn't use an incomplete path at all and when it is checked it is filled with the global incomplete path initially.
> > 
> > 
> > I believe this is now fixed - could you try this again?
> 
> Not fixed. Not sure if this matters when testing this: I can't add a torrent which has downloaded the metadata in this window anymore.

I was able to reproduce this. It only affects hybrid v1/v2 torrents.

The issue seems to stem from #21015. It appears to be related to `InfoHash`'s equality check (or maybe just to `TorrentDescriptor::parse`). When parsing the uploaded torrent file, we use the info hash as the key in `m_torrentMetadataCache`. At this stage, this info hash contains both the v1 and full v2 hash.

https://github.com/qbittorrent/qBittorrent/blob/71af105a897e3632e2e02fd6bfd85e30df3f7f81/src/webui/api/torrentscontroller.cpp#L2029-L2030

Then when checking `m_torrentMetadataCache` later, we perform a lookup using the InfoHash's `TorrentID`. With a hybrid torrent, the torrent id is the v2 hash truncated to 40 characters.

https://github.com/qbittorrent/qBittorrent/blob/71af105a897e3632e2e02fd6bfd85e30df3f7f81/src/webui/api/torrentscontroller.cpp#L1070-L1073

An `InfoHash` constructed from this truncated v2 hash is not considered equal to an `InfoHash` containing both the v1 and full v2 info hashes. I've attached a torrent that can be used to produce this. (@glassez @Chocobo1).

[torrent-many-files.torrent.zip](https://github.com/user-attachments/files/20874646/torrent-many-files.torrent.zip)

I can modify the new WebAPI routes or client behavior to work around this, for ex by using the full v2 hash, however I suspect this may be a broader issue we want to address. I think `InfoHash` should probably handle this directly without exposing v1/v2 compatibility issues to the caller.

Piccirello commented:
> > The add torrent files table has an effective minimum width of ~455px, which is what you're experiencing.
> 
> Guess i'm not explaining it well. [1.webm](https://github.com/user-attachments/assets/f43d83f2-db7d-4c6b-a30f-f2ed5784a894) Grab the edge Shrink window (to the right in this instance), it doesn't shrink. At that point i let go of mouse click. Move pointer to the left outside the window. When the contents of the window disappear during resize, i'm not holding down the button. I have to click again to get out of resize mode.
> 
> I didn't test it in mobile at all.

I don't believe this behavior is unique to these two new windows. I think it may be all mochaui windows loaded via iframe. For ex, I can reproduce the behavior of having to click again to get out of resize mode for the New Category window. I believe this may just be a mochaui quirk. I'm happy to fix this is we're able to identify something unique about this window though.

> > Can you retest with a non-hybrid torrent?
> 
> I didn't test hybrid or v2 at all. I can't find any legal torrents that contain a subfolder and i can't rename files and folders in this window so i can't provide a video. Best i can do is a diagram: Root folder ---File1 ---File2 ---File3 ---Folder ------File4 ------File5 ------File6
> 
> Select `Don't create subfolder`, and add the torrent. The Contents tab now shows this (`Folder` is now the root folder and files 1-3 don't show up): Folder ---File4 ---File5 ---File6

It's hard for me to reproduce without a relevant torrent file, though this sounds more likely to be a bug in the existing `torrents/add` endpoint, rather than in this PR. The Add Torrent window doesn't alter any behavior based on the selected Content layout - it just sends it to the API. Do you experience the same behavior with the previous add torrent dialog?

If this behavior is unique to this PR, could you share the payload from the POST to `torrents/add` when using the new add torrent window? The only thing I can think of is that the value of `filePriorities` is wrong, though I don't see how that would happen.

> > I was able to reproduce this. It only affects hybrid v1/v2 torrents.
> 
> I didn't test hybrid or v2 at all.

I'm not able to reproduce w/ a v1-only or v2-only torrent. After selecting those .torrent files, I'm able to add them without issue. The only torrent file I'm not able to add is a hybrid torrent. I'm also able to add a v1 torrent via magnet link, including after the metadata has been fetched.

Piccirello commented:
Pushed up some additional changes to achieve a pretty good experience with the new Add Torrent window on mobile. Will be putting up some other small PRs to fix small annoyances I've been running into.

@qbittorrent/bug-handlers could someone more focused on qBittorrent core check out the last quoted section of [this reply](https://github.com/qbittorrent/qBittorrent/pull/21645#issuecomment-2998572476). I believe there's a bug in the equality check of `InfoHash` and want to make sure this doesn't get lost. Happy to open a separate issue if needed.

> Same problem in master...

@thalieht would you be open to creating a new issue to report this bug with "Content layout"? I also believe all of your reported bugs have been addressed, but lmk if I missed any.

glassez commented:
> @qbittorrent/bug-handlers could someone more focused on qBittorrent core check out the last quoted section of [this reply](https://github.com/qbittorrent/qBittorrent/pull/21645#issuecomment-2998572476). I believe there's a bug in the equality check of `InfoHash` and want to make sure this doesn't get lost.

There is no bug. In general, there is no one-to-one mapping between `InfoHash` and `TorrentID`. If you want to retrieve values by `TorrentID` key, you should index them by `TorrentID` key. Even so, there are some pitfalls to consider.

glassez commented:
> What are the pitfalls associated with that?

The pitfalls may be related to the fact that every time you have some hash that is not directly obtained from the metadata of the torrent (i.e. not directly calculated from the info section), you can't be sure what it points to. E.g. when you (qBittorrent) are requested to retrieve metadata by <some_sha1_hash> you can't know if it's an info-hash of v1 torrent, v1 info-hash of hybrid torrent or truncated v2 info-hash of v2/hybrid torrent. The situation is even worse with the hybrid (combined sha1+sha256) hash. Without metadata, you can't even know if it's valid (i.e. both hashes refer to the same info section).

thalieht commented:
>I also believe all of your reported bugs have been addressed, but lmk if I missed any.

>>>Use another path for incomplete torrent is unchecked but the global path is still used. In the GUI if that option is unchecked it doesn't use an incomplete path at all and when it is checked it is filled with the global incomplete path initially.
>>>
>>I believe this is now fixed - could you try this again?
>
>Not fixed.

thalieht commented:
>@thalieht would you be open to creating a new issue to report this bug with "Content layout"? 

#22921

Piccirello commented:
> > What are the pitfalls associated with that?
> 
> The pitfalls may be related to the fact that every time you have some hash that is not directly obtained from the metadata of the torrent (i.e. not directly calculated from the info section), you can't be sure what it points to. E.g. when you (qBittorrent) are requested to retrieve metadata by <some_sha1_hash> you can't know if it's an info-hash of v1 torrent, v1 info-hash of hybrid torrent or truncated v2 info-hash of v2/hybrid torrent. The situation is even worse with the hybrid (combined sha1+sha256) hash. Without metadata, you can't even know if it's valid (i.e. both hashes refer to the same info section).

You're right about this. The biggest concern for the `torrents/add` endpoint would be a truncated v2 hash colliding with a v1 hash. The risk of this seems extremely low, so I think it'd be safe to move forward with changing `QHash<BitTorrent::InfoHash, BitTorrent::TorrentDescriptor> m_torrentMetadataCache` to `QHash<BitTorrent::TorrentID, BitTorrent::TorrentDescriptor> m_torrentMetadataCache`. It's technically a breaking change, but this API is so recent (and not yet in a release) that surely no one is using it yet. I'll put up a PR to address this.

> > > > Use another path for incomplete torrent is unchecked but the global path is still used. In the GUI if that option is unchecked it doesn't use an incomplete path at all and when it is checked it is filled with the global incomplete path initially.
> > > 
> > > 
> > > I believe this is now fixed - could you try this again?
> > 
> > 
> > Not fixed.

Just tested this and I do see a behavior difference. I just want to confirm the issue though so that we're on the same page. The issue I see is that, when "Use another path for incomplete torrent" is checked, the wrong path is filled in (it fills the default save path rather than the default incomplete download path). Otherwise, the checked/unchecked behavior is the same between the WebUI and the GUI. Is that right?

thalieht commented:
>I just want to confirm the issue though so that we're on the same page. The issue I see is that, when "Use another path for incomplete torrent" is checked, the wrong path is filled in (it fills the default save path rather than the default incomplete download path).

I didn't notice that. Now it works as expected.

>Otherwise, the checked/unchecked behavior is the same between the WebUI and the GUI. Is that right?

No, in GUI when it is unchecked, the Save Path is used i.e. the global/default "Use another path for incomplete torrent" is _not_ used. no matter if the global/default "Use another path for incomplete torrent" is enabled or not.
In WebUI, in this PR, the global/default "Use another path for incomplete torrent" _is_ used when this setting is unchecked.

Piccirello commented:
> No, in GUI when it is unchecked, the Save Path is used i.e. the global/default "Use another path for incomplete torrent" is _not_ used. no matter if the global/default "Use another path for incomplete torrent" is enabled or not. In WebUI, in this PR, the global/default "Use another path for incomplete torrent" _is_ used when this setting is unchecked.

I'm not able to reproduce the exact issue, but I think I know what is happening. Can you double check if you have "Keep incomplete torrents in" checked in qBittorrent options? I think what is happening is that the Add Torrent window isn't yet reflecting this preference. I'll push up a fix.

Piccirello commented:
Put up #22926 to address the bug with downloading hybrid torrents. I was mistaken earlier, this is not a breaking change.

Piccirello commented:
You're doing an awesome job finding these edge cases. All this state is a real pain to manage.

> 1. In Manual TMM, "Keep incomplete torrents in" checkbox is not initialized as checked (it should be, also when changing from Automatic to Manual, since the global setting is also checked)

I intentionally didn't add this, since the TMM is set to manual, but I think it's a fair argument that it should. And I see that the GUI does this. Fixed.

> 2. In Automatic TMM, select a category with an explicitly disabled incomplete save path and then select the "Empty" category. "Keep incomplete torrents in" checkbox should become enabled (just cosmetic)

I thought I had tested every iteration here but great find. Fixed.

> When "Keep incomplete torrents in" in qBittorrent options is **un**checked: In this window:
> 
> 1. In Automatic TMM, "Keep incomplete torrents in" checkbox is always unchecked (just cosmetic). It should be enabled when a category has an explicitly enabled incomplete save path.

Fixed.

> General issue in this window: When "Keep incomplete torrents in" checkbox is unchecked, its field should be empty.

Fixed.

Piccirello commented:
Temporarily putting this PR back into draft as it relies on some changes in #22938. The changes remove the need for the Add Torrent window to fetch categories and tags. Instead, it can now access them directly via the parent window.

Piccirello commented:
#22938 has been merged and so I have taken this out of draft.

Chocobo1 commented on src/webui/www/private/scripts/addtorrent.js:
```diff
@@ -0,0 +1,336 @@
+
+    let loadMetadataTimer;
+    const loadMetadata = (sourceUrl = undefined) => {
+        if (sourceUrl)
```
```suggestion
        if (sourceUrl !== undefined)
```

Chocobo1 commented on src/webui/www/private/scripts/torrent-content.js:
```diff
@@ -0,0 +1,572 @@
+
+    const getAllChildren = (id, fileId) => {
+        const getChildFiles = (node) => {
+            rowIds.push(node.data.rowId);
+            fileIds.push(node.data.fileId);
```
`rowIds` and `fileIds` are unbounded when `getChildFiles` is declared. I would declare `rowIds` and `fileIds` before this function to avoid this situation. The old code wasn't like this.

Piccirello replied:
This is the same as the original code - just moved.

Chocobo1 replied:
>This is the same as the original code - just moved.

Doesn't look like the original code. It reorganized the code.
https://github.com/qbittorrent/qBittorrent/blob/ae8a6689dc1451ae5d8e7a717d5c0a0f1056cb9d/src/webui/www/private/scripts/prop-files.js#L64-L96

Chocobo1 commented on src/webui/www/private/scripts/torrent-content.js:
```diff
@@ -0,0 +1,572 @@
+        const rowIds = [node.data.rowId];
+        const fileIds = [node.data.fileId];
+
+        node.children.forEach((child) => {
+            getChildFiles(child);
+        });
```
The old code is really poorly written. Instead of letting `getChildFiles` take a root node and do all operations within it, it duplicates code unnecessarily to the outside.

Piccirello replied:
This PR doesn't make it worse, so I think improving the existing implementation is best left for a future PR.

Chocobo1 replied:
don't mind that, it was just some mumbling...

Chocobo1 commented on src/webui/www/private/scripts/torrent-content.js:
```diff
@@ -0,0 +1,572 @@
+            onFilePriorityChanged(fileIds, priority);
+
+        const ignore = (priority === FilePriority.Ignored);
+        ids.forEach((_id) => {
```
please rename `_id` to something else.

Chocobo1 commented on src/webui/www/private/scripts/torrent-content.js:
```diff
@@ -0,0 +1,572 @@
+        for (const sibling of siblings) {
+            switch (getCheckboxState(sibling.rowId)) {
+                case TriState.Checked:
+                    checkedCount++;
+                    break;
+                case TriState.Unchecked:
+                    uncheckedCount++;
+                    break;
+                case TriState.Partial:
+                    indeterminateCount++;
```
```suggestion
                    ++checkedCount;
                    break;
                case TriState.Unchecked:
                    ++uncheckedCount;
                    break;
                case TriState.Partial:
                    ++indeterminateCount;
```

Piccirello replied:
This code was moved and is unchanged. Will leave modifications to a separate PR

Chocobo1 replied:
>This code was moved and is unchanged.

Could you point out where was the code in master?

Piccirello replied:
I was mistaken, this is new code. Will change.

Chocobo1 commented on src/webui/www/private/scripts/torrent-content.js:
```diff
@@ -0,0 +1,572 @@
+        }
+
+        const currentCheckboxState = getCheckboxState(parent.rowId);
+        let desiredCheckboxState;
+        if ((indeterminateCount > 0) || ((checkedCount > 0) && (uncheckedCount > 0)))
+            desiredCheckboxState = TriState.Partial;
+        else if (checkedCount > 0)
+            desiredCheckboxState = TriState.Checked;
+        else
+            desiredCheckboxState = TriState.Unchecked;
```
```suggestion
        let desiredCheckboxState = TriState.Unchecked;
        if ((indeterminateCount > 0) || ((checkedCount > 0) && (uncheckedCount > 0)))
            desiredCheckboxState = TriState.Partial;
        else if (checkedCount > 0)
            desiredCheckboxState = TriState.Checked;
```

Piccirello replied:
Also moved code so will leave for a separate PR

Chocobo1 replied:
>Also moved code so will leave for a separate PR

Could you point out where was the code in master?

Piccirello replied:
Was mistaken here as well, this is new code. Will update.

Chocobo1 commented on src/webui/www/private/scripts/torrent-content.js:
```diff
@@ -0,0 +1,572 @@
+
+        // listen for changes to torrentFilesFilterInput
+        document.getElementById("torrentFilesFilterInput").addEventListener("input", (event) => {
+            clearTimeout(torrentFilesFilterInputTimer);
```
```suggestion
            clearTimeout(torrentFilesFilterInputTimer);
            torrentFilesFilterInputTimer = -1;
```

Chocobo1 commented on src/webui/www/private/scripts/addtorrent.js:
```diff
@@ -0,0 +1,336 @@
+    };
+
+    const metadataCompleted = (showDownloadButton = true) => {
+        clearTimeout(loadMetadataTimer);
```
```suggestion
        clearTimeout(loadMetadataTimer);
        loadMetadataTimer = -1;
```

Chocobo1 commented on src/webui/www/private/scripts/addtorrent.js:
```diff
@@ -0,0 +1,336 @@
+    };
+
+    const metadataFailed = () => {
+        clearTimeout(loadMetadataTimer);
```
```suggestion
        clearTimeout(loadMetadataTimer);
        loadMetadataTimer = -1;
```

glassez commented on src/webui/www/private/scripts/client.js:
```diff
@@ -128,6 +130,82 @@ window.qBittorrent.Client ??= (() => {
+                }
+            })
+            .catch((error) => {
+                alert("QBT_TR(Unable to parse response)QBT_TR[CONTEXT=HttpServer]");
```
Is it a good idea to provide `${error.toString()` part to translators?

Piccirello replied:
> Is it a good idea to provide `${error.toString()` part to translators?

I don't think so, I prefer the alert how it currently is.

Piccirello replied:
Will adjust

Piccirello replied:
Pushed up this change. FYI `error.toString()` automatically includes the `Error:` prefix in the string, so I've omitted that.

glassez commented on src/webui/www/private/scripts/dynamicTable.js:
```diff
@@ -2887,16 +2893,18 @@ window.qBittorrent.DynamicTable ??= (() => {
             // progress
-            this.columns["progress"].updateTd = function(td, row) {
-                const value = Number(this.getRowValue(row));
+            if (this.columns["progress"]) {
```
I suppose that in JS `if (var["field"])` isn't the same as `if (var["field"] !== undefined)` so you insist on the second one?

Chocobo1 replied:
>I suppose that in JS if (var["field"]) isn't the same as if (var["field"] !== undefined) so you insist on the second one?

All [Truthy values](https://developer.mozilla.org/en-US/docs/Glossary/Truthy) will pass the first `if` and I don't want that.

Piccirello commented on src/webui/www/private/scripts/client.js:
```diff
@@ -128,6 +130,82 @@ window.qBittorrent.Client ??= (() => {
         return showingLogViewer;
     };
 
+    const createAddTorrentWindow = (title, source, metadata = undefined) => {
```
As you noted, this is to signal that `metadata` is optional. I don't see any benefit to switching from `undefined` to `null` for the optional param.

Chocobo1 replied:
or comment it, for example:
```js
    // `metadata` is optional
    const createAddTorrentWindow = (title, source, metadata) => {
```

Piccirello replied:
Comments fall out of date and are generally less helpful than self documenting code. I think defining an optional with `= undefined` is completely sensible. Also, this is definitely not worth spending time on compared to shipping this feature;

Piccirello commented on src/webui/www/private/scripts/dynamicTable.js:
```diff
@@ -2887,16 +2893,18 @@ window.qBittorrent.DynamicTable ??= (() => {
             // progress
-            this.columns["progress"].updateTd = function(td, row) {
-                const value = Number(this.getRowValue(row));
+            if (this.columns["progress"]) {
```
It's a constant time lookup so will omit for now

Chocobo1 replied:
>It's a constant time lookup so will omit for now

Just want to note that even lookup is constant time (assuming cost of `1`). 10000 lookups still has 10000 costs.

Piccirello commented on src/webui/www/private/scripts/addtorrent.js:
```diff
@@ -0,0 +1,336 @@
+    let windowId = "";
+    let source = "";
+
+    const LocalPreferences = new window.qBittorrent.LocalPreferences.LocalPreferences();
```
This is effectively a global that's also used by `DynamicTable`, so I had to keep it capitalized.

Chocobo1 replied:
I think I should look into it later...

luzpaz commented:
Heroic effort here. Thanks for the work to all involved!

Piccirello commented:
Let's merge!

Chocobo1 commented:
@Piccirello 
Thank you!

HanabishiRecca commented:
Can not add a torrent file if its name contains any non-ASCII characters.
There seems to be an issue with `parseMetadata` response, its encoding is wrong.

<img width="260" height="60" alt="image" src="https://github.com/user-attachments/assets/c12426fa-d3e2-4632-aebb-f3ca2476a427" />

HanabishiRecca commented:
Torrent content does not matter, only the file name matters. Get any torrent file and rename it to `тест.torrent` for example.

Piccirello commented:
> Torrent content does not matter, only the file name matters. Get any torrent file and rename it to `тест.torrent` for example.

Fixed in #23080

USBhost commented:
Adding multiple torrents that need a different path than the set default is annoying now. Can we have the path be the one before it?

Piccirello commented:
> Adding multiple torrents that need a different path than the set default is annoying now. Can we have the path be the one before it?

I'm planning to add a "Remember last used save path" to the Add Torrent dialog, like the GUI has. Would that make this any easier?

USBhost commented:
> > Adding multiple torrents that need a different path than the set default is annoying now. Can we have the path be the one before it?
> 
> I'm planning to add a "Remember last used save path" to the Add Torrent dialog, like the GUI has. Would that make this any easier?

It sure would.

Piccirello commented:
It should be pretty easy to add, though I'm waiting for #23088 to get merged.

USBhost commented:
Also there seems to be a problem when the client is open for awhile. I can't any any torrents magnet or files. The preview opens fine and it pulls the metadata but clicking add does not add the torrent.
<img width="558" height="56" alt="image" src="https://github.com/user-attachments/assets/aa650177-940e-4028-afc1-42f4cc19aec8" />

USBhost commented:
The max torrent size seems to be 61 files.

neatgz commented:
First of all, thank you for the great work.

When I upgraded my qBittorrent (NAS Docker x64) from 4.6.7 to 5.2.3, I discovered an issue.

If I need to open 30 torrent files at once, with the same save path and all settings being identical,

In the WebUI of version 5.2.3, **30 dialog boxes will pop up, requiring me to handle them 30 times!** This also causes browser lag.

In the WebUI of version 4.6.7, **there was only one dialog box. I just needed to set the save path, click OK, and everything was done.**

So, **I would like to ask if it's possible to add an option?** When adding a large number of torrents at once, I only need one dialog box, set it and click once, just like the old version.

Thank you very much.

USBhost commented:
Yeah that's my greatest grief with the new system. I haven't updated because of that. You should open issue for it.

zotabee commented:
@neatgz @USBhost It's coming, probably in the next release, 5.2.4.

"[5.2.0] Webui option to use 5.1.4 download windows" : https://github.com/qbittorrent/qBittorrent/issues/24181
"WebUI: Add shared dialog for adding multiple torrents" - #24386

Eagleames commented:
I strongly second the issue with multi-torrent uploads being a major issue, and a signifigant downgrade of this update.

I also note that since 99% of the time these features are NOT used, I would strongly suggest that these options be on an ADVANCED tab rather than on the main tab. The ACCEPT button at the bottom means scrolling every single time on every normal interface.

[ADD ALL QUEUED TORRENTS]   [CANCEL]   [ADVANCED OPTIONS]
